pkg/auth/remote_authorizer.go

```diff
@@ -0,0 +1,199 @@
+)
+
+type remoteAuthorizer struct {
+	remoteAuthClient auth_pb.AuthorizerClient
+	scope            *structpb.Value
+
+	clock            clock.Clock
+	maximumCacheSize int
+
+	lock sync.Mutex
+	// The channel is closed when the response is ready. If the response is
+	// missing from cachedResponses, the request failed and should be retried.
+	pendingRequests map[RemoteAuthorizerCacheKey]<-chan struct{}
+	cachedResponses map[RemoteAuthorizerCacheKey]remoteAuthorizerResponse
+	evictionSet     eviction.Set[RemoteAuthorizerCacheKey]
+}
+
+// RemoteAuthorizerCacheKey is the key type for the cache inside
+// remoteAuthorizer.
+type RemoteAuthorizerCacheKey [sha256.Size]byte
+
+type remoteAuthorizerResponse struct {
+	expirationTime time.Time
+	err            error
+}
+
+// NewRemoteAuthorizer creates a new Authorizer which asks a remote gRPC
+// service for authorize response. The result from the remote service is
+// cached.
+func NewRemoteAuthorizer(
+	client grpc.ClientConnInterface,
+	scope *structpb.Value,
+	clock clock.Clock,
+	evictionSet eviction.Set[RemoteAuthorizerCacheKey],
+	maximumCacheSize int,
+) Authorizer {
+	return &remoteAuthorizer{
+		remoteAuthClient: auth_pb.NewAuthorizerClient(client),
+		scope:            scope,
+
+		clock:            clock,
+		maximumCacheSize: maximumCacheSize,
+
+		pendingRequests: make(map[RemoteAuthorizerCacheKey]<-chan struct{}),
+		cachedResponses: make(map[RemoteAuthorizerCacheKey]remoteAuthorizerResponse),
+		evictionSet:     evictionSet,
+	}
+}
+
+func (a *remoteAuthorizer) Authorize(ctx context.Context, instanceNames []digest.InstanceName) []error {
+	errs := make([]error, 0, len(instanceNames))
+	for _, instanceName := range instanceNames {
+		errs = append(errs, a.authorizeSingle(ctx, instanceName))
+	}
+	return errs
+}
+
+func (a *remoteAuthorizer) authorizeSingle(ctx context.Context, instanceName digest.InstanceName) error {
+	authenticationMetadata := AuthenticationMetadataFromContext(ctx)
+	request := &auth_pb.AuthorizeRequest{
+		AuthenticationMetadata: authenticationMetadata.GetFullProto(),
+		Scope:                  a.scope,
+		InstanceName:           instanceName.String(),
+	}
+	if a.maximumCacheSize == 0 {
+		// Cache disabled, skip request deduplication.
+		response, err := a.authorizeRemotely(ctx, request)
+		if err != nil {
+			return err
+		}
+		return response.err
+	}
+
+	requestBytes, err := proto.Marshal(request)
+	if err != nil {
+		return util.StatusWrapWithCode(err, codes.PermissionDenied, "Failed to marshal authorize request")
+	}
+	// Hash the request to use as a cache key to save memory.
+	requestKey := sha256.Sum256(requestBytes)
+
+	for {
+		now := a.clock.Now()
+		a.lock.Lock()
+		if response, ok := a.getAndTouchCacheEntry(requestKey); ok && response.expirationTime.After(now) {
+			a.lock.Unlock()
+			return response.err
+		}
+		// No valid cache entry available. Deduplicate requests.
+		if responseReady, ok := a.pendingRequests[requestKey]; ok {
```

Comment:
Tiny nit: for readability, I'd appreciate it that we wrote code like this:

```go
if foo {
    A
    // no return here
} else {
   B
   return
}
```
As this instead:
```go
if !foo {
   B
   return
}
A
```

This makes it more explicit that the `select { case <-responseReady: ... }` falls through and causes the loop to run again, as it'll now be closer to the bottom of the loop.

Reply:
Also changed
```
select {
case <-ctx.Done():
    return ctx.Err()
case <-wakeupChan:
    A
}
```
into
```
select {
case <-ctx.Done():
    return ctx.Err()
case <-wakeupChan:
}
A
```